### ISSUE TITLE
Testing API Centos 7

### ISSUE THREAD

**Issue by crd1985** [wazuh/wazuh#2506]
# Ruleset test

| Version | Revision | Branch |
| --- | --- | --- |
| 3.9.0 | 3900 | 3.9 |

## Installation

- [x] Install API in a cluster of two nodes. One of the nodes must be a custom directory install. All agents must report to the worker node.
- [x] Check API status is running in both nodes.

## Configuration

- [x] Custom user, password and https working.

### Certificates and HTTPS

- [x] Install certificates.
- [x] Run query using HTTPS.

### Test mocha

Required tests:

- Agents.
- Decoders.
- Manager.
- Rootcheck.
- Rules.
- Syscheck.
- Syscollector.
- Policy monitoring.

Checks:

- [x] Master node:

    - [x] Centos 7 / Python 3.

**Comment by druizz90** [comment]
# Testing update

#### Environment
- Master 3.9 (upgraded)
- Worker 3.9 (upgraded)
- API 3.9 in master (upgraded) 

## Manual tests master API (new features)

#### Upgrade `ossec.conf` in worker nodes

```bash
curl -u foo:bar -k -X POST -H "Content-type:application/xml" --data-binary @/home/vagrant/ossec.conf-worker "https://localhost:55000/cluster/worker-1/files?path=etc/ossec.conf&pretty"
{
   "error": 0,
   "data": "File updated successfully"
}
```

New `ossec.conf` changes the name of worker node to `node02` as expected.

#### Upgrade `ossec.conf` in master node

```bash
curl -u foo:bar -k -X POST -H "Content-type:application/xml" --data-binary @/home/vagrant/ossec.conf-master "https://localhost:55000/cluster/master/files?path=etc/ossec.conf&pretty"
{
   "error": 0,
   "data": "File updated successfully"
}
```

New `ossec.conf` changes the name of worker node to `node01` as expected.

#### Post files

Update `local_rules.xml`:

```bash
curl -u foo:bar -k -X POST -H "Content-type:application/xml" --data-binary @/home/vagrant/new_rules.xml "https://localhost:55000/cluster/node01/files?path=etc/rules/new_rules.xml&pretty"
{
   "error": 0,
   "data": "File updated successfully"
}
```

If new `ossec.conf` is corrupted:

```bash
curl -u foo:bar -X POST -H "Content-type:application/xml" --data-binary @/root/ossec.conf-corrupted "http://localhost:55000/cluster/node01/files?path=etc/ossec.conf&pretty"
{
   "error": 703,
   "message": "Invalid XML file. closing tag no</logall_json is expected inplace of global."
}
```
#### Get validation

Example with `ossec.conf` OK (master):

```bash
curl -u foo:bar -k -X GET -k "https://172.19.0.100:55000/cluster/node01/configuration/validation?pretty"
{
   "error": 0,
   "data": {
      "status": "OK"
   }
}
```

Example with `ossec.conf` corrupted (worker):

```bash
curl -u foo:bar -k -X GET -k "https://172.19.0.100:55000/cluster/node02/configuration/validation?pretty"
{
   "error": 0,
   "data": {
      "status": "KO",
      "details": [
           {
                "node": "worker-1",
                "error": "(1226): Error reading XML file '/var/ossec/etc/ossec.conf':  (line 0)."
           }
      ]
   }
}
```

Get validation from all nodes fails, only shows validation for first node. We should fix this:

```bash
curl -u foo:bar -k -X GET -k "https://172.19.0.100:55000/cluster/configuration/validation?pretty"
{
   "error": 0,
   "data": {
      "status": "OK"
   }
}
```

Validation with `local_rules.xml` corrupted:

```bash
curl -k -u foo:bar -k -X GET "https://127.0.0.1:55000/manager/configuration/validation?pretty"
{
   "error": 0,
   "data": {
      "status": "KO",
      "details": [
         "(1226): Error reading XML file 'etc/rules/local_rules.xml': XMLERR: Element '1.1.1.1</srcip' not closed. (line 17).",
         "(1220): Error loading the rules: 'etc/rules/local_rules.xml'."
      ]
   }
}
```

Validation with `ossec.conf` corrupted:

```bash
curl -k -u foo:bar -k -X GET "https://127.0.0.1:55000/manager/configuration/validation?pretty"
{
   "error": 3006,
   "message": "Error reading cluster configuration"
}
```
This call fails if cluster is disabled too with the same error in all API calls.

#### Get files

```bash
curl -u foo:bar -k -X GET "https://localhost:55000/cluster/node02/files?path=etc/ossec.conf&pretty"
{
   "error": 0,
   "data": "  <!--\n  Wazuh - Manager - Default configuration.\n  More info at: https://documentation.wazuh.com\n  Mailing list: https://groups.google.com/forum/#!forum/wazuh\n-->\n  <ossec_config>\n    <global>\n      <jsonout_output><yes</jsonout_output>\n      <alerts_log>yes</alerts_log>\n      <logall>no</logall>\n      <logall_json>no</logall_json>\n      <email_notification>no</email_notification>\n      <smtp_server>smtp.example.wazuh.com</smtp_server>\n      <email_from>ossecm@example.wazuh.com</email_from>\n      <email_to>recipient@example.wazuh.com</email_to>\n      <email_maxperhour>12</email_maxperhour>\n      <email_log_source>alerts.log</email_log_source>\n    </global>\n    <cluster>\n      <name>wazuh</name>\n      <node_name>node02</node_name>\n      <key>9d273b53510fef702b54a92e9cffc82e</key>\n      <node_type>worker</node_type>\n      <port>1516</port>\n      <bind_addr>172.19.0.100</bind_addr>\n      <nodes>\n        <node>172.19.0.100</node>\n      </nodes>\n      <hidden>no</hidden>\n      <disabled>no</disabled>\n    </cluster>\n    <!-- Choose between plain or json format (or both) for internal logs -->\n    <logging>\n      <log_format>plain</log_format>\n    </logging>\n    <alerts>\n      <log_alert_level>3</log_alert_level>\n      <email_alert_level>12</email_alert_level>\n    </alerts>\n    <remote>\n      <connection>secure</connection>\n      <port>1514</port>\n      <protocol>udp</protocol>\n    </remote>\n    <!-- Policy monitoring -->\n    <rootcheck>\n      <disabled>no</disabled>\n      <!-- Frequency that rootcheck is executed - every 12 hours -->\n      <frequency>43200</frequency>\n      <rootkit_files>/var/ossec/etc/shared/rootkit_files.txt</rootkit_files>\n      <rootkit_trojans>/var/ossec/etc/shared/rootkit_trojans.txt</rootkit_trojans>\n      <system_audit>/var/ossec/etc/shared/system_audit_rcl.txt</system_audit>\n      <system_audit>/var/ossec/etc/shared/system_audit_ssh.txt</system_audit>\n      <system_audit>/var/ossec/etc/shared/cis_debian_linux_rcl.txt</system_audit>\n      <skip_nfs>yes</skip_nfs>\n    </rootcheck>\n    <wodle name=\"open-scap\">\n      <disabled>yes</disabled>\n      <timeout>1800</timeout>\n      <interval>1d</interval>\n      <scan-on-start>yes</scan-on-start>\n      <content path=\"ssg-debian-8-ds.xml\" type=\"xccdf\">\n        <profile>xccdf_org.ssgproject.content_profile_common</profile>\n      </content>\n      <content path=\"cve-debian-oval.xml\" type=\"oval\"/>\n    </wodle>\n    <wodle name=\"syscollector\">\n      <disabled>no</disabled>\n      <interval>1h</interval>\n      <scan_on_start>yes</scan_on_start>\n      <hardware>yes</hardware>\n      <os>yes</os>\n      <network>yes</network>\n    </wodle>\n    <wodle name=\"vulnerability-detector\">\n      <disabled>no</disabled>\n      <interval>1d</interval>\n      <run_on_start>yes</run_on_start>\n      <update_ubuntu_oval interval=\"60m\" version=\"16,14,12\">yes</update_ubuntu_oval>\n      <update_redhat_oval interval=\"60m\" version=\"7,6,5\">yes</update_redhat_oval>\n    </wodle>\n    <wodle name=\"key-request\">\n      <enabled>yes</enabled>\n      <timeout>60</timeout>\n      <script>my_script.sh</script>\n      <threads>4</threads>\n      <queue_size>1024</queue_size>\n    </wodle>\n    <!-- File integrity monitoring -->\n    <syscheck>\n      <disabled>no</disabled>\n      <!-- Frequency that syscheck is executed default every 12 hours -->\n      <frequency>43200</frequency>\n      <scan_on_start>yes</scan_on_start>\n      <!-- Generate alert when new file detected -->\n      <alert_new_files>yes</alert_new_files>\n      <!-- Don't ignore files that change more than 3 times -->\n      <auto_ignore>no</auto_ignore>\n      <!-- Directories to check  (perform all possible verifications) -->\n      <directories check_all=\"yes\">/etc,/usr/bin,/usr/sbin</directories>\n      <directories check_all=\"yes\">/bin,/sbin,/boot</directories>\n      <!-- Files/directories to ignore -->\n      <ignore>/etc/mtab</ignore>\n      <ignore>/etc/hosts.deny</ignore>\n      <ignore>/etc/mail/statistics</ignore>\n      <ignore>/etc/random-seed</ignore>\n      <ignore>/etc/random.seed</ignore>\n      <ignore>/etc/adjtime</ignore>\n      <ignore>/etc/httpd/logs</ignore>\n      <ignore>/etc/utmpx</ignore>\n      <ignore>/etc/wtmpx</ignore>\n      <ignore>/etc/cups/certs</ignore>\n      <ignore>/etc/dumpdates</ignore>\n      <ignore>/etc/svc/volatile</ignore>\n      <ignore>/sys/kernel/security</ignore>\n      <ignore>/sys/kernel/debug</ignore>\n      <!-- File types to ignore -->\n      <ignore type=\"sregex\">.log$|.swp$</ignore>\n      <!-- Check the file, but never compute the diff -->\n      <nodiff>/etc/ssl/private.key</nodiff>\n      <skip_nfs>yes</skip_nfs>\n    </syscheck>\n    <!-- Active response -->\n    <global>\n      <white_list>127.0.0.1</white_list>\n      <white_list>^localhost.localdomain$</white_list>\n      <white_list>10.0.0.2</white_list>\n    </global>\n    <command>\n      <name>disable-account</name>\n      <executable>disable-account.sh</executable>\n      <expect>user</expect>\n      <timeout_allowed>yes</timeout_allowed>\n    </command>\n    <command>\n      <name>restart-ossec</name>\n      <executable>restart-ossec.sh</executable>\n      <expect/>\n    </command>\n    <command>\n      <name>firewall-drop</name>\n      <executable>firewall-drop.sh</executable>\n      <expect>srcip</expect>\n      <timeout_allowed>yes</timeout_allowed>\n    </command>\n    <command>\n      <name>host-deny</name>\n      <executable>host-deny.sh</executable>\n      <expect>srcip</expect>\n      <timeout_allowed>yes</timeout_allowed>\n    </command>\n    <command>\n      <name>route-null</name>\n      <executable>route-null.sh</executable>\n      <expect>srcip</expect>\n      <timeout_allowed>yes</timeout_allowed>\n    </command>\n    <command>\n      <name>win_route-null</name>\n      <executable>route-null.cmd</executable>\n      <expect>srcip</expect>\n      <timeout_allowed>yes</timeout_allowed>\n    </command>\n    <!--\n  <active-response>\n    active-response options here\n  </active-response>\n  -->\n    <!-- Log analysis -->\n    <localfile>\n      <log_format>syslog</log_format>\n      <location>/var/ossec/logs/active-responses.log</location>\n    </localfile>\n    <localfile>\n      <log_format>syslog</log_format>\n      <location>/var/log/messages</location>\n    </localfile>\n    <localfile>\n      <log_format>syslog</log_format>\n      <location>/var/log/auth.log</location>\n    </localfile>\n    <localfile>\n      <log_format>syslog</log_format>\n      <location>/var/log/syslog</location>\n    </localfile>\n    <localfile>\n      <log_format>command</log_format>\n      <command>df -P</command>\n      <frequency>360</frequency>\n    </localfile>\n    <localfile>\n      <log_format>full_command</log_format>\n      <command>netstat -tan |grep LISTEN |grep -v 127.0.0.1 | sort</command>\n      <frequency>360</frequency>\n    </localfile>\n    <localfile>\n      <log_format>full_command</log_format>\n      <command>last -n 5</command>\n      <frequency>360</frequency>\n    </localfile>\n    <ruleset>\n      <!-- Default ruleset -->\n      <decoder_dir>ruleset/decoders</decoder_dir>\n      <rule_dir>ruleset/rules</rule_dir>\n      <rule_exclude>0215-policy_rules.xml</rule_exclude>\n      <list>etc/lists/audit-keys</list>\n      <!-- User-defined ruleset -->\n      <decoder_dir>etc/decoders</decoder_dir>\n      <rule_dir>etc/rules</rule_dir>\n    </ruleset>\n    <!-- Configuration for ossec-authd\n       To enable this service, run:\n       ossec-control enable auth\n  -->\n    <auth>\n      <disabled>no</disabled>\n      <port>1515</port>\n      <use_source_ip>yes</use_source_ip>\n      <force_insert>yes</force_insert>\n      <force_time>0</force_time>\n      <purge>yes</purge>\n      <use_password>no</use_password>\n      <!-- <ssl_agent_ca></ssl_agent_ca> -->\n      <ssl_verify_host>no</ssl_verify_host>\n      <ssl_manager_cert>/var/ossec/etc/sslmanager.cert</ssl_manager_cert>\n      <ssl_manager_key>/var/ossec/etc/sslmanager.key</ssl_manager_key>\n      <ssl_auto_negotiate>no</ssl_auto_negotiate>\n    </auth>\n    <agentless>\n      <type>ssh_integrity_check_linux</type>\n      <frequency>300</frequency>\n      <host>admin@192.168.1.108</host>\n      <state>periodic_diff</state>\n      <arguments>/etc /usr/bin /usr/sbin</arguments>\n    </agentless>\n    <active-response>\n      <disabled>no</disabled>\n      <command>host-deny</command>\n      <location>defined-agent</location>\n      <agent_id>032</agent_id>\n      <level>10</level>\n      <rules_group>sshd,|pci_dss_11.4,</rules_group>\n      <timeout>1</timeout>\n    </active-response>\n    <syslog_output>\n      <level>9</level>\n      <server>192.168.1.241</server>\n    </syslog_output>\n    <integration>\n      <name>virustotal</name>\n      <api_key>API_KEY</api_key>\n      <!-- Replace with your VirusTotal API key -->\n      <group>syscheck</group>\n      <alert_format>json</alert_format>\n      <hook_url/>\n    </integration>\n    <socket>\n      <name>custom_socket</name>\n      <location>/var/run/custom.sock</location>\n      <mode>tcp</mode>\n      <prefix>custom_syslog: </prefix>\n    </socket>\n    <global>\n      <email_notification>yes</email_notification>\n      <email_to>hello@wazuh.com</email_to>\n      <smtp_server>localhost</smtp_server>\n      <email_from>wazuh@test.com</email_from>\n    </global>\n    <email_alerts>\n      <email_to>you@example.com</email_to>\n      <level>4</level>\n      <do_not_delay/>\n    </email_alerts>\n  </ossec_config>\n"
}
```

```bash
curl -u foo:bar -k -X GET "https://localhost:55000/cluster/node02/files?path=etc/rules/local_rules.xml&pretty"
{
   "error": 0,
   "data": "<!-- Local rules -->\n\n<!-- Modify it at your will. -->\n<!-- Copyright (C) 2015-2019, Wazuh Inc. -->\n\n<!-- Example -->\n<group name=\"local,syslog,sshd,\">\n\n  <!--\n  Dec 10 01:02:02 host sshd[1234]: Failed none for root from 1.1.1.1 port 1066 ssh2\n  -->\n  <rule id=\"100001\" level=\"5\">\n    <if_sid>5716</if_sid>\n    <srcip>1.1.1.1</srcip>\n    <description>sshd: authentication failed from IP 1.1.1.1.</description>\n    <group>authentication_failed,pci_dss_10.2.4,pci_dss_10.2.5,</group>\n  </rule>\n\n</group>\n"
}
```

```bash
curl -u foo:bar -k -X GET "https://localhost:55000/cluster/node02/files?path=etc/lists/audit-keys&pretty"
{
   "error": 0,
   "data": "audit-wazuh-w:write\naudit-wazuh-r:read\naudit-wazuh-a:attribute\naudit-wazuh-x:execute\naudit-wazuh-c:command\n"
}
```

#### Restart manager

If `ossec.conf` is corrupted:

```bash
curl -u foo:bar -k -X PUT "https://127.0.0.1:55000/cluster/node02/restart?pretty"
{
   "error": 0,
   "data": "Restarting manager"
}

-----------------------
2019/02/14 11:04:28 ossec-testrule: CRITICAL: (1226): Error reading XML file '/var/ossec/etc/ossec.conf':  (line 0).
```

If `ossec.conf` is OK:
```bash
curl -u foo:bar -k -X PUT "https://127.0.0.1:55000/cluster/node02/restart?pretty"
{
   "error": 0,
   "data": "Restarting manager"
}

---------------------
2019/02/14 11:06:46 ossec-monitord: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2019/02/14 11:06:46 ossec-logcollector: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2019/02/14 11:06:47 ossec-remoted: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2019/02/14 11:06:47 ossec-syscheckd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2019/02/14 11:06:47 ossec-analysisd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2019/02/14 11:06:47 ossec-execd: INFO: (1314): Shutdown received. Deleting responses.
2019/02/14 11:06:47 ossec-execd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2019/02/14 11:06:47 wazuh-db: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
```

Restart master node using API from worker:

```bash
curl -u foo:bar -k -X PUT "https://172.19.0.100:55000/cluster/node01/restart?pretty"
{
   "error": 0,
   "data": "Restarting manager"
}
---------------------
Restart OK
```
I found errors when all cluster nodes are restarted:

```bash
curl -u foo:bar -k -X PUT "https://127.0.0.1:55000/cluster/restart?pretty"
{
   "error": 0,
   "data": "Restarting manager"
}
---------------------
node02 restarted
node01 not restarted

```

```bash
curl -u foo:bar -k -X PUT "https://172.19.0.100:55000/cluster/restart?pretty"
{
   "error": 0,
   "data": "Restarting manager"
}
---------------------
node02 restarted
node01 not restarted
```

**Comment by druizz90** [comment]
Errors found in this issue were solved in #2602.
